### ISSUE TITLE
Ghost hack causes crashes on Usamune 1.93 U

### ISSUE THREAD

**Issue by krithalith** [FramePerfection/STROOP#4]
I am using the Usamune 1.93 U ROM, on Mupen 1.1.8-2, and STROOP refactor 0.6.1.

I reset the ROM and loaded an empty save file. After the intro cutscene and text, I activated the ghost hack, made a state, and loaded the state. Everything seems fine. The white flying Mario ghosts appears and circles around me. However, from this point, certain Mario actions / animations will make the game crash.

No problem:
idle, walking, turning around, jump, jumpkick, double jump, crouch, crouch slide, breakdance, ...

Crash:
backflip, longjump, sideflip, triple jump, dive, ...

Mupen crash log (longjump):

Mupen 64 1.1.8-2
27/08/2024 21:55:13
Address: 0x000000 (closest: mupen64(2).exe 0x000000)
Stack overflow (0xc0000005)
Video: C:\Users\Derek\Desktop\Mupen64-RR_Repack22\Mupen64-RR Repack\plugin\glN64.dll
Audio: C:\Users\Derek\Desktop\Mupen64-RR_Repack22\Mupen64-RR Repack\plugin\a_Jabo_Dsound.dll
Input: C:\Users\Derek\Desktop\Mupen64-RR_Repack22\Mupen64-RR Repack\plugin\TASInput.dll
RSP: C:\Users\Derek\Desktop\Mupen64-RR_Repack22\Mupen64-RR Repack\plugin\hacktarux-azimer-rsp-hle.dll
VCR Task: 0
Emu Launched: 1

**Comment by FramePerfection** [comment]
Thanks for the detailed description.

I'll leave the [Usamune](https://sites.google.com/view/supermario64/usamunepj/usamunerom) link here for myself, as I fear that I'll have to do some assembly debugging here... 😰

**Comment by FramePerfection** [comment]
Okay, so I've tried replicating the crash with a longjump, a triple jump and a backflip, on 0.6.2 (or rather, the current development state, which isn't much different from the latest nightly build), and it worked just fine.  
I've found that indeed some changes were made to the ghost hack since 0.6.1 (I wonder who made those, might have been me :suspect: ), so this might work for you as well.

**Comment by krithalith** [comment]
Ah yep 0.6.2 works without issues with Usamune (U) 1.93. Thanks!

**Comment by krithalith** [comment]
Ah upon further testing it only seems to work if you don't use any of the actual Usamune features like showing RTA times etc, also if you open the usamune menu it still crashes. I'm guessing the additional HUD elements and the custom menu also use the Gfx pool in ways the vanilla game does not. Speficially I found the following crashes:

If Usamune misc RTA timing is already active and the ghost hack is activated, the game will crash on certain actions like decribed previously.

If Usamune is in pure vanilla (no usamune features active yet), the ghost hack can be activated and seems to work for all actions, but the game will crash when pressing D-down to access the Usamune menu.

I see that in 0.6.2 there is an option to "move" the gfx pool. Is there any chance this could solve the issue?

**Comment by FramePerfection** [comment]
> I see that in 0.6.2 there is an option to "move" the gfx pool. Is there any chance this could solve the issue?

No, that can't be it. The GFX pool just collects valid commands, and so long as Usamune and the ghost hack are producing valid commands (which they do), there is not going to be a conflict there.  
Moving the GFX pool is intended only to allow for more ghosts, because each ghost instance produces a considerable amount of commands that, when significantly multiplied, need more space than the original game has accounted for.

It is more likely that _activating_ Usamune feature will cause memory clashes. I will try to figure out the details behind the crashes you've been observing, since I haven't actually attempted to use Usamune myself in regards to the issue (since you had initially described a reproduction scenario that didn't end up causing it, but could have).

Anyways, good to know. I had hoped that this issue had poofed out of existance with 0.6.2, but apparently that is not really the case...

**Comment by hugou74130** [comment]
### Root cause: Memory collision with Usamune's expanded RAM. Still reproduces on the latest Mupen version.

I reproduced this reliably and traced it with a debugger. The crash is caused by a corrupted execution flow resulting from both the ghost hack and Usamune writing into the same expanded RAM region.

### Reproduction
ROM: Usamune v1.93u (8 MB / Expansion Pak)

### Steps
Enable the ghost hack (the hovering placeholder ghost appears), then open the Usamune menu using the D-pad (certain Mario actions also trigger it when the matching Usamune options are active).

### Differential test
 With the ghost hack disabled, everything works fine. The ghost hack is a necessary ingredient — this is an interaction.

### Reproduced on two setups
Older Mupen: Hard crash, c0000005 access violation.

### Latest Mupen (1.4.0-6)
The newer core detects the issue and stops with the message "A critical emulation error has occurred: Executed RESERVED" / "unusual execution pattern". Same root cause, but handled gracefully.

### Debugger Analysis (WinDbg, on the hard-crash setup)
c0000005, Parameter[0] = 8 → an execute access violation (jump to non-executable memory), eip = 0x04001000.

The faulting instruction is mupen's block dispatcher: eax = jump_table[target >> 16]; call eax. At the time of the crash, ecx = 0x0000FFFF, meaning the N64 jump target was corrupted to 0xFFFFxxxx, indexing an uninitialized table slot (0x04001000) and calling into garbage data.

The current precomp_instr has the N64 address 0x8033AF5C — inside the game's data region (just below gMarioStates at 0x8033B170), with operand fields set to 0xBAADF00D (uninitialized memory poison). Execution had already derailed into data.

On the latest core, this exact derailment is what triggers the "Executed RESERVED" message (the CPU decoded data as if it were an instruction).

### Likely Root Cause
The ghost hack claims a fixed slice of expanded RAM: GhostTab.cs writes control bytes at addresses 0x80407F90–0x80408000 (including 0xFF at 0x80407FFC), zeroes out roughly 0x80407000–0x80408000, and writes ghost data starting from 0x80409B00+. The injected loop resides at 0x80408000 (ghost_hack_US.asm) and animation buffers are at 0x8050xxxx.

Usamune also uses expanded RAM for its menu and features. The regions overlap, causing the ghost hack to overwrite Usamune's data → Usamune later reads a corrupted pointer → execution derails → crash / RESERVED. This also explains why disabling the hack resolves the issue.

### Suggested Fixes
Detect & Warn: 
Before claiming the region, verify that it is actually empty (filled with zeroes). If a ROM hack already occupies it, display a warning or refuse the operation instead of silently corrupting memory. This is a simple, safe change that turns a silent failure into a clear message.

Relocatable Region:
Make the ghost hack's base address configurable (similar to the existing "move GFX pool" feature) so it can reside in free RAM depending on the ROM hack. This is the ideal fix for proper coexistence, but it is more complex (requires removing hardcoded addresses in ghost_loop.asm).

Note: I did not perform the final hardware watchpoint step to capture the exact overwritten pointer; therefore, the collision is strongly evidenced but not confirmed down to the exact byte.
